core/break.lua

```diff
@@ -74,34 +75,60 @@ function lineBreak:trimGlue() -- 842
   nodes[#nodes+1] = SILE.nodefactory.penalty(inf_bad)
 end
 
+function lineBreak:parShape()
+  return { width = self.hsize, left = 0, right = 0 }
+  -- NOTE FOR DEVELOPERS: this method is called when the linebreak.parShape
+  -- setting is true.
+  -- The default implementation does nothing fancy and just outputs regular
+  -- paragraphs (at the cost of an extra function call).
+  -- Extended paragraph shapes are intended to be provided by overriding
+  -- this method.
+  --
+  -- The expected return value is a table:
+  --    { width=..., left=..., right=... }
+  -- defining the width, left indent and right indent for the given line.
+  --
+  -- The width must be a SILE.measurement()'s, and the indents absolute
+  -- numbers.
+  --
+  -- (Obviously, one should ensure left + right + width = self.hsize, so
+  -- returning the two indents could have been avoided, but it feels more
+  -- readable and natural this way, and easier afterwards back into the
+  -- typesetter).
+  --
+  -- Using a Lua function rather than a fixed array allows one to perform
+  -- computations, but remember the linebreak methods will be called multiple
+  -- times to try good breaks, so it is strongly recommended to "memoize"
```

Comment:
Additionally I would like to build the memorization into the breaker so that the end user supplied function doesn't have to optimize this, we will just know not to keep asking the same question.

Reply:
I'm not strongly opinionated on the issue, but I'm not fully sure about it, for the following reasons at least:
- If using a pre-computed  shape (as done in the test), additional memoization (https://en.wikipedia.org/wiki/Memoization) is of course not needed (= would cost extra memory usage for no reason).
- If computations are kept simple (e.g. a mere triangle/trapezoid), the small hit on performances might still be acceptable vs. caching values.
- If one wants to implement more complex algorithms, e.g. shaping several paragraphs at once in a common shape, it may require additional things to be overloaded (in the typesetter instance, etc.) to keep tracks, etc. The breaker (as it currently stands) might be too low level then anyway. I've not explored this topic much, however (what I have been exploring a bit, though, without a satisfying result yet, is shapes with one hole... Until I abandoned, it was a bit frustrating and for needs that might not be that relevant - I kind of know that LaTeX has the `shapepar` package which fully reimplements Knuth's parshaping differently - with apparently edge cases too though - but I am not sure it is that widely used, beyond demonstration...)

Reply:
I have implemented the API changes I suggested. I'm adding a caching layer onto it now, will post for review soon.